viamonvif/mdns_server.go

```diff
@@ -0,0 +1,93 @@
+package viamonvif
+
+import (
+	"net"
+	"sync"
+
+	"github.com/edaniels/zeroconf"
+	"go.viam.com/rdk/logging"
+)
+
+type mdnsInfo struct {
+	ip      net.IP
+	cleanup func()
+}
+
+type mdnsServer struct {
+	mu            sync.Mutex
+	mappedDevices map[string]mdnsInfo
+	logger        logging.Logger
+}
+
+func newMDNSServer(logger logging.Logger) mdnsServer {
+	return mdnsServer{
+		mappedDevices: make(map[string]mdnsInfo),
+		logger:        logger,
+	}
+}
+
+// serialNumber is expected to be a device serial number. Without any `.local` TLD suffix. The IP is
+// the IP to be mapped to. An error is returned if the mdns mapping failed. Informing the caller
+// that doing an ip -> hostname substitution is not expected to work.
+func (server *mdnsServer) Add(serialNumber string, ip net.IP) error {
+	server.mu.Lock()
+	defer server.mu.Unlock()
+
+	if info, exists := server.mappedDevices[serialNumber]; exists {
+		if info.ip.Equal(ip) {
+			// We're already managing an mdns entry for the same hostname -> ip mapping.
+			return nil
+		}
+
+		info.cleanup()
+	}
+	delete(server.mappedDevices, serialNumber)
+
+	avoidLintMagicNumberComplaintThatIsntEvenRelevant := 8080
+	mdnsServer, err := zeroconf.RegisterProxy(
+		serialNumber, // Dan: As far as I can tell, just a name.
+		"_rtsp._tcp", // Dan: The mDNS "service" to register. Doesn't make a difference?
```

Comment:
It is possible for the communication between viamrtsp and the camera to be over udp.  Will that cause problems with this `.tcp` name?

Reply:
I didn't see any problems in testing regarding tcp/udp. Strictly speaking, ping is _neither_ tcp nor udp. Nor is ping seeking the rtsp service. And ping does* work/stop working as this RegisterProxy/"mdnsServer.Shutdown" methods are called.

This is admittedly pretty magic. Thus the documentation on every input of what I found to be necessary and how it maps to the side-effects a user can see. I'm happy to run a specific experiment.

More broadly, there's buy-in from product that if this just doesn't seem to be working, we'll try a different approach such as editing /etc/hosts. Or using "macros" or "templates" such that the viam-server does the IP substitution before passing a configuration to camera construction/reconfigure.